### PR TITLE
[MRG] Make result data type of KDE evaluation like in the input data (Issue #11464)

### PR DESCRIPTION
**Pull request by lschwetlick**
As discussed in Issue #11464, currently the KDE evaluation always returns float64 arrays, regardless of the point input data type. This PR makes the data type agree with the input datatype, as long as it is a floating datatype. If it's not a float, I'm not sure this change would make any sense.

#### Reference issue
Closes gh-#11464.

#### What does this implement/fix?
Checks the input data type and instantiate the results array with this dtype if it floats.

#### Additional information
I can't get scipy to build right now, that's why I'm putting it as work in progress.

### CODE REVIEW

**Comment by lschwetlick** [comment]
The failing checks are for 32 bit systems that don't have a float128 datatype. Windows
Would it be bad practice to skip these tests on systems that dont have it? I've seen something like this used [here](https://github.com/scipy/scipy/blob/master/scipy/linalg/tests/test_decomp.py#L2633)

```
@pytest.mark.skipif(np.dtype(np.intp).itemsize < 8, reason="dont test on 32-bit, because float128 doesnt exist there")
```

but I'm not that familiar with the internals of why some systems have 128 floats and others dont.

**Comment by mdhaber** [comment]
> Would it be bad practice to skip these tests on systems that dont have it?

That sounds reasonable to me, but I'll let someone else chime in. Thanks for the patch!

> I can't get scipy to build right now, that's why I'm putting it as work in progress.

What OS are you using? Are you running into an error?

**Comment by lschwetlick** [comment]
> What OS are you using? Are you running into an error?

It seems to be some kind of path problem or venv problem. I'm on MacOs Catalina. When I do `pip install -e .` in the scipy repo, in my virtual env I get a 11 918 line error message  that starts with
```
Running setup.py develop for scipy
    ERROR: Command errored out with exit status 1:
```
And then just seems to be missing a bunch of libraries
```
libraries mkl_rt not found in ['/virtual_envs/scipy_env/lib', '/usr/local/lib', '/usr/lib']
libraries openblas not found 
libraries openblas,lapack not found 
etc.
```

**Comment by rgommers** [comment]
Yes, skipping tests when `float128` (or `float96`) doesn't exist is the right thing to do. This is probably the cleaner way to spell it (from `test_mstats_basic.py`):
```
@pytest.mark.skipif(not hasattr(np, 'float96'), reason='cannot find float96 so skipping')
```

**Comment by lschwetlick** [comment]
I don't understand why the CI is failing for python 3.6. The errors do't seem to be related to my changes. Can someone help?

**Comment by mdhaber** [comment]
@lschwetlick you're right, those aren't your fault and they are occurring on all PRs right now. occasionally when there are upstream changes we get CI failures that take a while to fix.

**Review by mdhaber — APPROVED** [review]
Looks good to me! I'm not sure I understand the bit about parameterizing with integers instead of the data types directly, so I'll wait for @rgommers to review again since he's already commented.

**Comment by lschwetlick** [comment]
> parameterizing with integers instead of the data types directly
I realize it's kind of a hack, but the problem was:
The parametrize decorator gets evaluated despite the skipif decorator. Aparently they are not hierarchical. So, If I parametrize with the data types directly, on systems that don't have that data type it evaluates and throws an error within the parametrize decorator.

**Comment by lschwetlick** [comment]
> Regarding build issues, have you tried the [Quickstart Guide](https://scipy.github.io/devdocs/dev/contributor/quickstart_mac.html#quickstart-mac)?

Thanks, I hadn't! So far I've only ever used pip and I didn't want to risk making more of a mess by mix-and-matching. I definitely need to figure it out though! Thanks for your help :)

File: scipy/stats/tests/test_kdeoth.py
Comment (mdhaber) [inline scipy/stats/tests/test_kdeoth.py]: How about:
```suggestion
@pytest.mark.parametrize("test_type", [getattr(np, dtype, None) for dtype in ['float32', 'float64', 'float96', 'float128']])
```

Comment (lschwetlick) [reply scipy/stats/tests/test_kdeoth.py]: That sounds good. I'll do it like that!

**Comment by mdhaber** [comment]
@WarrenWeckesser do you have a minute to take a look at this? LGTM, so please merge if you agree.

**Comment by lschwetlick** [comment]
Does anything else need to be done on this?

**Comment by mdhaber** [comment]
Thanks for your patience @lschwetlick. I am confident that this PR does what it was intended to do. I talked with @WarrenWeckesser, though, and he had some thoughts that the desired behavior might be a bit more subtle. I think he will post his comments soon.

**Comment by WarrenWeckesser** [comment]
@lschwetlick, a belated "thanks!" for contributing this PR.  Sorry for not commenting sooner.

With this change, the data type of the result of evaluating the object will match the data type of the input argument when the input is a floating point type.

Typical case:

    dataset = [... some data, with some data type ...]
    f = gaussian_kde(dataset)

    points = np.array([... some points ...], dtype=np.float32)
    y = f(points)

In master, `y` has dtype `float64`.  With the change in this pull request, `y` has dtype `float32` (matches the data type of `points`).

This looks like reasonable behavior, but I'm not sure it is the best choice.

Consider the full calculation performed by the sequence

    f = gaussian_kde(dataset, bw_method, weights)
    y = f(points)

The result `y` depends on not just `points`, but also on the original inputs `dataset`, `bw_method` (which can be a floating point scalar) and `weights`.  We can think of the calculation as the result of a function `F` that takes all those arguments

    y = F(points, dataset, bw_method, weights)

If, for example, `dataset` and `weights` have data type `float128`, we would probably expect the usual numpy casting rules to apply, so the result `y` should also have type `float128`, even if `dataset` has type `float32`.

So perhaps we really need the equivalent to

    output_dtype = np.common_type(points, dataset, bw_width, weight)

(See [common_type](https://numpy.org/devdocs/reference/generated/numpy.common_type.html).) We can't use that line literally, because `bw_width` might be a
string, `weights` might be None, etc.  Presumably there is a subset of the attributes of the `gaussian_kde` object that would be appropriate arguments to `np.common_type(points, ...)`, e.g. `self.inv_cov`, `self.factor`, and others.

I haven't traced the code paths from the original inputs to the attributes to figure this out, but I think that would determine an appropriate set of attributes to include in a call to `np.common_type`.  It would probably be something like

    output_dtype = np.common_type(points, self.inv_cov, self.factor, self.weights)

but I haven't checked that carefully.

File: scipy/stats/kde.py
Comment (WarrenWeckesser) [inline scipy/stats/kde.py]: `np.float` is just an alias for the Python built-in `float`, and the use of that alias is discouraged.  I think most numpy devs would agree that exposing that name in the `numpy` namespace was a mistake.

Instead of `np.float`, use `np.float64`, which is the actual numpy data type that we want.

Comment (lschwetlick) [reply scipy/stats/kde.py]: thats good to know, I'd wondered at that before!

**Comment by lschwetlick** [comment]
Thanks for the input, what you say makes a lot of sense. I'll try to implement that!

**Comment by lschwetlick** [comment]
Just to document my thought process concerning 
```
output_dtype = np.common_type(self.covariance, points)
```
[Line 573](https://github.com/lschwetlick/scipy/blob/e1bbfbe8c8970b8f9bdb2efb1f929e6e30842fc1/scipy/stats/kde.py#L573)
 uses `self.weights` and `self.dataset`, so we have their common type  in `self._data_covariance`.
[Line 578](https://github.com/lschwetlick/scipy/blob/e1bbfbe8c8970b8f9bdb2efb1f929e6e30842fc1/scipy/stats/kde.py#L578) uses `self.factor` (bandwidth) and `self._data_covariance`, so the common type of weights, dataset and bandwith is in `self.covariance`.
Therefore, in `evaluate` we only need to check points and covariance.

File: scipy/stats/tests/test_kdeoth.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_kdeoth.py]: The code in these decorators could be simplified by determining the floating point types once,  and filtering out the undefined types instead of substituting `None`, i.e.:

```
_ftypes = [getattr(np, dtype) for dtype in ['float32', 'float64',
                                            'float96', 'float128']
           if hasattr(np, dtype)]
```

Then use `_ftypes` in the decorators, instead of repeating the list comprehension in each one.

File: scipy/stats/tests/test_kdeoth.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_kdeoth.py]: Remove space at the end of the line.

File: scipy/stats/tests/test_kdeoth.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_kdeoth.py]: If you use the suggestion I gave for computing `_ftypes`, these checks for `None` can be removed.

File: scipy/stats/tests/test_kdeoth.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_kdeoth.py]: PEP-8: Two blank lines before `def`.

File: scipy/stats/tests/test_kdeoth.py
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_kdeoth.py]: As before, don't using `np.int`.  In fact, it looks like you can remove this line, and just use `points = np.arange(5)` below.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Thanks @lschwetlick, looks good.   I suggested a few changes in the unit tests.

File: scipy/stats/tests/test_kdeoth.py
Comment (lschwetlick) [reply scipy/stats/tests/test_kdeoth.py]: Yes, thats nicer!

File: scipy/stats/tests/test_kdeoth.py
Comment (lschwetlick) [reply scipy/stats/tests/test_kdeoth.py]: In fact, I think that test can just be combined with the other one, now that we're using the common type. That should cover integer datatypes in the same way as floats!

**Comment by lschwetlick** [comment]
Well... looks like I was wrong about
> Line 573 uses self.weights and self.dataset, so we have their common type in self._data_covariance.

Turns out numpy's `cov` does not return the datatype expected by `common_type`. It always gives back a minimum of float64, and only goes to float128 when the data is float128. The weights don't matter.
```
weights = np.arange(5, dtype=np.float128)
dataset = np.arange(5, dtype=np.float64)
data_covariance = cov(dataset, rowvar=1,
                                   bias=False,
                                   aweights=weights)
```
gives
```
print(data_covariance.dtype) -> float64
print(dataset.dtype) -> float64
print(weights.dtype) -> float128
print(np.common_type(weights, dataset)) -> <class 'numpy.float128'>
```

**Comment by lschwetlick** [comment]
So I decided to handle it in the same way as numpy's `cov` since that made the most sense to me, i.e. discount the weight datatype. If you agree, feel free to merge!

**Comment by WarrenWeckesser** [comment]
Merged. Thanks for your patience and persistence, @lschwetlick.  I hope you'll be interested in contributing more bug fixes and enhancements--there is plenty to do!

**Pull request merged**
